openwisp_controller/connection/api/serializer.py

```diff
@@ -33,3 +38,51 @@ class Meta:
             'created',
             'modified',
         ]
+
+
+class CredentialSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    params = serializers.JSONField()
+
+    class Meta:
+        model = Credentials
+        fields = (
+            'id',
+            'connector',
+            'name',
+            'organization',
+            'auto_add',
+            'params',
+            'created',
+            'modified',
+        )
+        read_only_fields = ('created', 'modified')
+
+
+class DeviceConnectionSerializer(
+    FilterSerializerByOrgManaged, ValidatedModelSerializer
+):
+    class Meta:
+        model = DeviceConnection
+        fields = (
+            'id',
+            'credentials',
+            'update_strategy',
+            'enabled',
+            'is_working',
+            'failure_reason',
```

Comment:
Shouldn't reason of failure be read-only?

Reply:
similarly to the conclusion we came in the config API, if we avoid these kind of restrictions in the REST API we can allow users to build whatever they want with it.